### PR TITLE
Add ReferencesArray and DynamicReferencesArray helpers

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

**Comment by vjik** [comment]
This is helpers for references from factory. May be move them to `yiisoft/factory`?

**Comment by yiiliveext** [comment]
> This is helpers for references from factory. May be move them to `yiisoft/factory`?

It is used in the container configs only

**Comment by samdark** [comment]
Need usage examples in phpdoc. Now it's not clear what's the benefit in having these.

**Comment by yiiliveext** [comment]
> Need usage examples in phpdoc. Now it's not clear what's the benefit in having these.

See https://github.com/yiisoft/data-response/pull/60

File: src/DynamicReferencesArray.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+use Yiisoft\Factory\Exception\InvalidConfigException;
+
+final class DynamicReferencesArray
+{
+    /**
+     * @param array $ids
+     * @return array
+     * @throws InvalidConfigException
+     */
+    static function from(array $ids)
```
Comment (rustamwin) [inline src/DynamicReferencesArray.php:17]:
```suggestion
    public static function from(array $ids)
```

File: src/ReferencesArray.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+use Yiisoft\Factory\Exception\InvalidConfigException;
+
+final class ReferencesArray
+{
+    /**
+     * @param array $ids
+     * @return array
+     * @throws InvalidConfigException
+     */
+    static function from(array $ids)
```
Comment (rustamwin) [inline src/ReferencesArray.php:17]:
```suggestion
    public static function from(array $ids)
```

**Review by rustamwin — APPROVED** [review]

File: src/ReferencesArray.php (hunk 2 above)
Comment (yiiliveext) [reply src/ReferencesArray.php]: Fixed

File: src/DynamicReferencesArray.php (hunk 1 above)
Comment (yiiliveext) [reply src/DynamicReferencesArray.php]: Fixed

File: src/DynamicReferencesArray.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+use Yiisoft\Factory\Exception\InvalidConfigException;
+
+final class DynamicReferencesArray
+{
+    /**
+     * @param array $ids
+     * @return array
+     * @throws InvalidConfigException
+     */
+    public static function from(array $ids)
```
Comment (vjik) [inline src/DynamicReferencesArray.php:17]:
```suggestion
    /**
     * @param string[] $ids
     * @return DynamicReference[]
     * @throws InvalidConfigException
     */
    public static function from(array $ids): array
```

File: src/DynamicReferencesArray.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        $references = [];
+
+        foreach ($ids as $key => $id) {
+            if (!is_string($id)) {
```
Comment (vjik) [inline src/DynamicReferencesArray.php:22]: Add `use is_string` for more performance.

File: src/ReferencesArray.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+use Yiisoft\Factory\Exception\InvalidConfigException;
+
+final class ReferencesArray
+{
+    /**
+     * @param array $ids
+     * @return array
+     * @throws InvalidConfigException
+     */
+    public static function from(array $ids)
```
Comment (vjik) [inline src/ReferencesArray.php:17]:
```suggestion
    /**
     * @param string[] $ids
     * @return Reference[]
     * @throws InvalidConfigException
     */
    public static function from(array $ids): array
```

File: src/ReferencesArray.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        $references = [];
+
+        foreach ($ids as $key => $id) {
+            if (!is_string($id)) {
```
Comment (vjik) [inline src/ReferencesArray.php:22]: Add `use is_string` for more performance.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/ReferencesArray.php (hunk 6 above)
Comment (yiiliveext) [reply src/ReferencesArray.php:69]: Fixed

File: src/ReferencesArray.php (hunk 5 above)
Comment (yiiliveext) [reply src/ReferencesArray.php]: Fixed

File: src/DynamicReferencesArray.php (hunk 4 above)
Comment (yiiliveext) [reply src/DynamicReferencesArray.php:69]: Fixed

File: src/DynamicReferencesArray.php (hunk 3 above)
Comment (yiiliveext) [reply src/DynamicReferencesArray.php]: Fixed

**Pull request merged**